src/liblelantus/lelantus_prover.cpp

```diff
@@ -101,32 +102,70 @@ void LelantusProver::generate_sigma_proofs(
     a.resize(N);
     std::vector<Scalar> serialNumbers;
     serialNumbers.reserve(N);
-    for (std::size_t i = 0; i < N; ++i)
-    {
-        if (!c.count(Cin[i].second))
-            throw std::invalid_argument("No such anonymity set or id is not correct");
-
-        GroupElement gs = (params->get_g() * Cin[i].first.getSerialNumber().negate());
-        serialNumbers.emplace_back(Cin[i].first.getSerialNumber());
-        std::vector<GroupElement> C_;
-        C_.reserve(c.size());
-
-        const auto& set = c.find(Cin[i].second);
-        if (set == c.end())
-            throw std::invalid_argument("No such anonymity set");
-
-        for (auto const &coin : set->second)
-            C_.emplace_back(coin.getValue() + gs);
-
-        rA[i].randomize();
-        rB[i].randomize();
-        rC[i].randomize();
-        rD[i].randomize();
-        Tk[i].resize(params->get_sigma_m());
-        Pk[i].resize(params->get_sigma_m());
-        Yk[i].resize(params->get_sigma_m());
-        a[i].resize(params->get_sigma_n() * params->get_sigma_m());
-        sigmaProver.sigma_commit(C_, indexes[i], rA[i], rB[i], rC[i], rD[i], a[i], Tk[i], Pk[i], Yk[i], sigma[i], sigma_proofs[i]);
+
+    std::size_t threadsMaxCount = std::min((unsigned int)N, std::thread::hardware_concurrency());
+    std::vector<std::shared_future<void>> parallelTasks;
+    parallelTasks.reserve(threadsMaxCount);
+    ParallelOpThreadPool<void> threadPool(threadsMaxCount);
+
+    std::vector<std::vector<GroupElement>> C_;
```

Comment:
you can store only one array, and clear() it at the start of each iteration

Reply:
We are passing all values by reference, storing and clearing will not work.